### PR TITLE
create/delete/update the irule profile and bwc policy dynamically whe…

### PR DESCRIPTION
**Pull request by jikui**
…n creating/deleting a loadbalancer

@<reviewer_id>
#### What issues does this address?
Fixes #<issueid>
WIP #<issueid>
...

### CODE REVIEW

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 1]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+                    if irule_name in vs_payload['rules']:
+                        vs_payload['rules'].remove(irule_name)
+                self.vs_helper.update(bigip, vs_payload)
+        return
+
+    def __add_bwc(self, bigip, loadbalancer, service, update_listener=True):
+        # The logic is , add the irule bwc policies and go through listeners to
+        # bind the bwc policy.
+        bandwidth = loadbalancer.get('bandwidth')
+        bandwidth = bandwidth * 1000 * 1000
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: 1024 or 1000 ?

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 2]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+        return bwc_profile
+
+    def __get_bwc_profile_name(self, loadbalancer):
+
+        profile_name = '/' + self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id']) + \
+            '/' + 'bwc_' + 'profile_' + loadbalancer['id']
+        return profile_name
+
+    def __get_irule_profile_name(self, loadbalancer):
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: This terminology may looks confused. iRule, profile and bwc policy are different things, and you are create an iRule whose content has a reference to a bwc policy. Maybe we can name it a little better. Such kind of confusing name also exists in some other places.

**Review by zhaoqin-github — COMMENTED** [review]
Does your code works when loadbalancer model does not include bandwidth parameter? Because China mobile private cloud and public cloud are different things, we hope not to break anything, if this option is not there.

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py (hunk 1 above)
Comment (jikui) [reply f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: qin, I tested that if we use 1024, the value configured in bigip is not correct. So I changed to 1000 and it worked well.

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py (hunk 2 above)
Comment (jikui) [reply f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: accept.

**Comment by jikui** [comment]
> Does your code works when loadbalancer model does not include bandwidth parameter? Because China mobile private cloud and public cloud are different things, we hope not to break anything, if this option is not there.
It should be working fine while I didn't get a test-bed for the testing.

File: f5_openstack_agent/lbaasv2/drivers/bigip/service_adapter.py [hunk 3]
```diff
@@ -482,11 +482,6 @@ def _map_virtual(self, loadbalancer, listener,
         self._add_vlan_and_snat(listener, vip)
         self._add_profiles_session_persistence(listener, pool, vip)
 
-        # Set bandwidth controller from loadbalancer's setting.
-        # applying order:
-        #   loadbalancer's bandwidth field -> l7policy's name(bwc name)
-        self._add_listener_bwcPolicy(loadbalancer, vip)
-
         # Set bandwidth controller
         if 'bwcPolicy' in listener:
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/service_adapter.py:486]: Should we also remove these lines?

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 4]
```diff
@@ -306,6 +487,18 @@ def __init__(self, driver):
             "connection_limit": "connectionLimit"
         }
 
+    def __get_bwc_profile_name(self, loadbalancer):
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy  ?

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 5]
```diff
@@ -306,6 +487,18 @@ def __init__(self, driver):
...
+        profile_name = '/' + self.driver.service_adapter.\
+                       get_folder_name(loadbalancer['tenant_id']) + \
+                       '/bwc_profile_' + loadbalancer['id']
+        return profile_name
+
+    def __get_irule_profile_name(self, loadbalancer):
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: get_irule_profile  -->  get_bwc_irule  ?

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 6]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+    def __delete_bwc(self, bigip, loadbalancer, service):
+        # The logic is , unbond the listenres' bwc policy and delete the irule
+        # bwc policies and go through listeners to
+
+        # go through the listeners and delete them
+        self.__update_listener_bwc(bigip, loadbalancer, service, False)
+        payload = self._create_irule_payload(loadbalancer, True)
+        super(LoadBalancerManager, self)._delete(bigip, payload,
+                                                 None, None,
+                                                 type="irule-profile",
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule-profile --> irule  ?

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 7]
```diff
@@ -192,6 +192,12 @@ def __init__(self, driver):
             resource_helper.ResourceType.virtual_address)
         self.tenant_manager = tenants.BigipTenantManager(self.driver.conf,
                                                          self.driver)
+        self.irule_helper = resource_helper.BigIPResourceHelper(
+            resource_helper.ResourceType.rule)
+        self.bwc_profile_helper = resource_helper.BigIPResourceHelper(
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile  -->  bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 8]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+            payload = self._create_bwc_payload(loadbalancer, 0)
+            super(LoadBalancerManager, self).\
+                _delete(bigip, payload, None, None, type="bwc-profile",
+                        helper=self.bwc_profile_helper)
+
+        super(LoadBalancerManager, self)._delete(bigip, payload,
+                                                 loadbalancer, service)
+
+    def _create_bwc_payload(self, loadbalancer, bandwidth):
+        # create irule profile
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule profile  -->  bwc irule  ?

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 9]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+        bwc_profile = {}
+        bwc_profile['partition'] = self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id'])
+        bwc_profile['name'] = 'bwc_profile_' + loadbalancer['id']
+        if bandwidth > 0:
+            bwc_profile['maxRate'] = bandwidth * 1000 * 1000
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 10]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+            super(LoadBalancerManager, self).\
+                _delete(bigip, payload, None, None, type="bwc-profile",
+                        helper=self.bwc_profile_helper)
+
+        super(LoadBalancerManager, self)._delete(bigip, payload,
+                                                 loadbalancer, service)
+
+    def _create_bwc_payload(self, loadbalancer, bandwidth):
+        # create irule profile
+        bwc_profile = {}
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 11]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+        bwc_profile['partition'] = self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id'])
+        bwc_profile['name'] = 'bwc_profile_' + loadbalancer['id']
+        if bandwidth > 0:
+            bwc_profile['maxRate'] = bandwidth * 1000 * 1000
+        return bwc_profile
+
+    def __get_bwc_profile_name(self, loadbalancer):
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 12]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+
+        profile_name = '/' + self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id']) + \
+            '/bwc_profile_' + loadbalancer['id']
+        return profile_name
+
+    def __get_irule_profile_name(self, loadbalancer):
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> bwc_irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 13]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+        profile_name = '/' + self.driver.service_adapter.\
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: profile_name --> irule_name

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 14]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+        return profile_name
+
+    def _create_irule_payload(self, loadbalancer, meta_only=False):
+        # create irule profile
+        irule_profile = {}
+        irule_profile['partition'] = self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id'])
+        irule_profile['name'] = 'irule_bwc_' + loadbalancer['id']
+        if not meta_only:
+            bwc_profile_name = self.__get_bwc_profile_name(loadbalancer)
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 15]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+
+    def __get_irule_profile_name(self, loadbalancer):
+        profile_name = '/' + self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id']) + '/irule_bwc_'\
+            + loadbalancer['id']
+        return profile_name
+
+    def _create_irule_payload(self, loadbalancer, meta_only=False):
+        # create irule profile
+        irule_profile = {}
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 16]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+            '/bwc_profile_' + loadbalancer['id']
+        return profile_name
+
+    def __get_irule_profile_name(self, loadbalancer):
+        profile_name = '/' + self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id']) + '/irule_bwc_'\
+            + loadbalancer['id']
+        return profile_name
+
+    def _create_irule_payload(self, loadbalancer, meta_only=False):
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule --> bwc_irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 17]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+
+        # create irule
+        if 'bandwidth' in loadbalancer.keys() and \
+           loadbalancer['bandwidth']:
+            LOG.debug("bandwidth exists")
+            # return
+            payload = self._create_irule_payload(loadbalancer, True)
+            super(LoadBalancerManager, self).\
+                _delete(bigip, payload, None, None,
+                        type="irule-profile",
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule-profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 18]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+                        helper=self.irule_helper)
+
+            payload = self._create_bwc_payload(loadbalancer, 0)
+            super(LoadBalancerManager, self).\
+                _delete(bigip, payload, None, None, type="bwc-profile",
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc-profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 19]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+        bwc_profile['name'] = 'bwc_profile_' + loadbalancer['id']
+        if bandwidth > 0:
+            bwc_profile['maxRate'] = bandwidth * 1000 * 1000
+        return bwc_profile
+
+    def __get_bwc_profile_name(self, loadbalancer):
+
+        profile_name = '/' + self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id']) + \
+            '/bwc_profile_' + loadbalancer['id']
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 20]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+            get_folder_name(loadbalancer['tenant_id']) + '/irule_bwc_'\
+            + loadbalancer['id']
+        return profile_name
+
+    def _create_irule_payload(self, loadbalancer, meta_only=False):
+        # create irule profile
+        irule_profile = {}
+        irule_profile['partition'] = self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id'])
+        irule_profile['name'] = 'irule_bwc_' + loadbalancer['id']
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_bwc_ --> bwc_irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 21]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+    def __get_irule_profile_name(self, loadbalancer):
+        profile_name = '/' + self.driver.service_adapter.\
+            get_folder_name(loadbalancer['tenant_id']) + '/irule_bwc_'\
+            + loadbalancer['id']
+        return profile_name
+
+    def _create_irule_payload(self, loadbalancer, meta_only=False):
+        # create irule profile
+        irule_profile = {}
+        irule_profile['partition'] = self.driver.service_adapter.\
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py (hunk 20 above)
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 22]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+            get_folder_name(loadbalancer['tenant_id'])
+        irule_profile['name'] = 'irule_bwc_' + loadbalancer['id']
+        if not meta_only:
+            bwc_profile_name = self.__get_bwc_profile_name(loadbalancer)
+            irule_profile['apiAnonymous'] = "when SERVER_CONNECTED {\n\
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 23]
```diff
@@ -212,6 +218,72 @@ def create(self, loadbalancer, service, **kwargs):
...
+                    BWC::policy attach " + bwc_profile_name + "\n}"
+        return irule_profile
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 24]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+                if 'bandwidth' in loadbalancer.keys() and \
+                   loadbalancer['bandwidth']:
+                    return 'Add'
+
+        return 'None'
+
+    def __update_listener_bwc(self, bigip, loadbalancer, service, add=True):
+
+        if 'listeners' in service.keys():
+            irule_name = self.__get_irule_profile_name(loadbalancer)
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> bwc_irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 25]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+                self.vs_helper.update(bigip, vs_payload)
+        return
+
+    def __add_bwc(self, bigip, loadbalancer, service, update_listener=True):
+        # The logic is , add the irule bwc policies and go through listeners to
+        # bind the bwc policy.
+        bandwidth = loadbalancer.get('bandwidth')
+        payload = self._create_bwc_payload(loadbalancer, bandwidth)
+        super(LoadBalancerManager, self)._create(
+            bigip, payload, None, None, type="bwc-profile",
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc-profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 26]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+            helper=self.bwc_profile_helper, overwrite=True)
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc-profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 27]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+
+        payload = self._create_irule_payload(loadbalancer)
+        super(LoadBalancerManager, self)._create(
+            bigip, payload, None, None, type="irule-profile",
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule-profile --> irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 28]
```diff
@@ -513,6 +706,11 @@ def _create(self, bigip, vs, listener, service):
         if listener['protocol'] == "HTTP" or \
            listener['protocol'] == "TERMINATED_HTTPS":
             self._create_http_profile(bigip, vs)
+        if 'bandwidth' in loadbalancer.keys() and \
+           loadbalancer['bandwidth']:
+            LOG.debug("bandwidth exists")
+            irule_name = self.__get_irule_profile_name(loadbalancer)
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: irule_profile --> bwc_irule

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 29]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+        payload = self._create_irule_payload(loadbalancer, True)
+        super(LoadBalancerManager, self)._delete(bigip, payload,
+                                                 None, None,
+                                                 type="irule-profile",
+                                                 helper=self.irule_helper)
+
+        payload = self._create_bwc_payload(loadbalancer, 0)
+        super(LoadBalancerManager, self).\
+            _delete(bigip, payload, None, None,
+                    type="bwc-profile", helper=self.bwc_profile_helper)
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 30]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+        return
+
+    def __modify_bwc(self, bigip, loadbalancer):
+        # The logic is quit simple with updating the bandwidth
+        bandwidth = loadbalancer.get('bandwidth')
+        payload = self._create_bwc_payload(loadbalancer, bandwidth)
+        super(LoadBalancerManager, self).\
+            _create(bigip, payload, None, None,
+                    type="bwc-profile", helper=self.bwc_profile_helper,
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py]: bwc_profile --> bwc_policy

File: f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py [hunk 31]
```diff
@@ -233,9 +305,118 @@ def _pre_create(self, service):
...
+                    overwrite=True)
+
+    def _update_bwc(self, old_loadbalancer, loadbalancer, service):
+        operate = self.__get_update_operation(old_loadbalancer, loadbalancer)
```
Comment (zhaoqin-github) [inline f5_openstack_agent/lbaasv2/drivers/bigip/resource_manager.py:375]: __get_update_operation() looks complex, but I feel the logic is not really complex. Do you think the following code looks correct?

```
if old_loadbalancer.get("bandwidth", 0) == loadbalancer.get("bandwidth", 0):
    # Do nothing
else:
    loadbalancer.get("bandwidth", 0) > 0:
        # add_bwc(), which create/update bwc policy and bwc irule, ensure bwc irule is attached to every VS
    loadbalancer.get("bandwidth", 0) <= 0:
        # delete_bwc(), which detach bwc irule from every VS, delete bwc irule and bwc policy (optional)
```

**Review by zhaoqin-github — CHANGES_REQUESTED** [review]

File: f5_openstack_agent/lbaasv2/drivers/bigip/service_adapter.py (hunk 3 above)
Comment (jikui) [reply f5_openstack_agent/lbaasv2/drivers/bigip/service_adapter.py:486]: qin, 

485 and 485 were introduced by 304583a3 which we didn't remove for this time.

stack@devstack-agent-lite:~/f5lbaas-agent$ git show 304583a3
commit 304583a3bb6ec65563a1878fdc6c73136498b2e7
Author: Qin Zhao <q.zhao@f5.com>
Date:   Thu Feb 27 16:56:20 2020 +0800

    Support bandwidth control of listener

Comment (zhaoqin-github) [reply f5_openstack_agent/lbaasv2/drivers/bigip/service_adapter.py:486]: Understand. That is listener level bwc setting. We may re-think it and remove the code in the future.

**Review by zhaoqin-github — APPROVED** [review]
Most of the code looks good. Let us test them carefully.

**Pull request merged**